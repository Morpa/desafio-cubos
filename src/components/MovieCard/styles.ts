import styled, { css } from 'styled-components'
import { motion } from 'framer-motion'
import media from 'styled-media-query'

import * as MovieTypeStyles from 'components/MovieTypes/styles'

export const Wrapper = styled(motion.div)`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    flex: 1;
    border-radius: ${theme.border.radius};
  `}
`
export const MovieWrapper = styled(motion.a)`
  ${({ theme }) => css`
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;
    align-items: stretch;
    align-content: center;
    margin: 2rem 0;
    width: 100%;
    height: 32rem;
    margin-bottom: 4rem;
    background: ${theme.colors.cardBackground};
    text-decoration: none;
    color: ${theme.colors.textColor};

    ${media.lessThan('medium')`
      flex-direction: column-reverse;
      width: 100%;
      height: 100%;
    `}
  `}
`

export const Cover = styled.img`
  ${({ theme }) => css`
    width: 20rem;
    height: 100%;
    background: ${theme.colors.primary};
    object-fit: cover;

    ${media.lessThan('medium')`
        width: 95%;
        margin: 0 auto;
        max-height: 100%;
    `}
  `}
`

export const Content = styled.div`
  flex: 1;
  flex-direction: column;
  display: block;
`

export const Header = styled.h1`
  ${({ theme }) => css`
    display: flex;
    align-items: flex-end;
    justify-items: flex-start;
    height: 6.4rem;
    flex: 1;
    padding-left: 9.6rem;
    padding-right: ${theme.spacings.xxsmall};
    background: ${theme.colors.primary};
    color: ${theme.colors.acqua};
    overflow: auto;
    white-space: nowrap;
    font-family: 'Abel';

    ${media.lessThan('medium')`
    align-items: center;
        white-space: normal;
        min-height:90%;
        font-size:${theme.font.sizes.medium}
      `};
  `}
`

export const InfosWrapper = styled.div`
  ${({ theme }) => css`
    flex: 1;
    flex-direction: column;
    padding: ${theme.spacings.small};
    overflow: hidden;
  `}
`

export const Date = styled.p`
  ${({ theme }) => css`
    align-self: flex-start;
    font-size: ${theme.font.sizes.large};
    color: ${theme.colors.textColor};
  `}
`

export const Overview = styled.div`
  flex: 1;
  max-height: 13rem;
  overflow: hidden;
  text-overflow: clip;
`

export const Categories = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: flex-start;
    margin-top: ${theme.spacings.small};
    padding: ${theme.spacings.xxsmall} 0;

    ${media.lessThan('medium')`
      flex-wrap: wrap;
      justify-content: center;
      margin-top: ${theme.spacings.xxsmall};

      ${MovieTypeStyles.Wrapper} {
        margin: 0.4rem;
      }
    `}
  `}

  ${MovieTypeStyles.Wrapper} {
    margin-right: 0.4rem;
  }
`

export const BadgeWrapper = styled.div`
  ${({ theme }) => css`
    width: 7rem;
    height: 7rem;
    position: absolute;
    margin-top: ${theme.spacings.small};
    margin-left: ${theme.spacings.xsmall};

    ${media.lessThan('medium')`
      margin-top: ${theme.spacings.small};
    `}
  `}
`
