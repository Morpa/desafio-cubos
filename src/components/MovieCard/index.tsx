import Link from 'next/link'
import { MotionProps } from 'framer-motion'

import Badge from 'components/Badge'
import MovieTypes from 'components/MovieTypes'

import floatToPercentage from 'utils/floatToPercentage'
import formatToBR from 'utils/formatToBR'

import * as S from './styles'

export type MovieCardProps = {
  poster_path: string
  overview: string
  release_date: string
  genre_ids: number[]
  title: string
  vote_average: number
  layoutId?: string
}

const animationSpring: MotionProps = {
  whileHover: { y: -4 },
  whileTap: {
    y: 0
  },
  transition: {
    type: 'spring',
    stiffness: 250,
    damping: 14
  }
}

const MovieCard = ({
  poster_path,
  overview,
  release_date,
  genre_ids,
  title,
  vote_average,
  layoutId
}: MovieCardProps) => (
  <S.Wrapper {...animationSpring}>
    <Link href="/" passHref>
      <S.MovieWrapper>
        <S.Cover
          src={`https://image.tmdb.org/t/p/w342/${poster_path}`}
          alt={`Poster of the movie ${title}`}
          layoutId={layoutId}
        />

        <S.Content>
          <S.BadgeWrapper>
            <Badge>{floatToPercentage(vote_average)}%</Badge>
          </S.BadgeWrapper>
          <S.Header>{title}</S.Header>
          <S.InfosWrapper>
            <S.Date>{formatToBR(release_date)}</S.Date>
            <S.Overview>{overview}</S.Overview>
            <S.Categories>
              {genre_ids.map((genre) => (
                <MovieTypes key={genre} name="Suspense" />
              ))}
            </S.Categories>
          </S.InfosWrapper>
        </S.Content>
      </S.MovieWrapper>
    </Link>
  </S.Wrapper>
)

export default MovieCard
