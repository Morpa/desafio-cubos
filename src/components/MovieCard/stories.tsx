import { Story, Meta } from '@storybook/react/types-6-0'
import MovieCard, { MovieCardProps } from '.'

export default {
  title: 'MovieCard',
  component: MovieCard,
  args: {
    genre_ids: [35, 10751, 14, 10749],
    original_title: 'Xuxa - Lua de Cristal',
    overview:
      'Maria da Graça é uma jovem bonita e sonhadora que se muda para a cidade grande com a intenção de fazer aulas de canto. Lá, ela se hospeda na casa de sua tia Zuleika e seus primos Lidinha e Mauricinho, que vivem atormentando sua vida, fazendo-a trabalhar como uma escrava. Por ser um tanto ingênua e tímida, Maria vive caindo nas armações de Lidinha; enquanto é alvo das constantes cantadas de Mauricinho, ao que ela rejeita; tendo de suportar também os detestáveis amigos deste, enquanto não consegue se adaptar á cidade. Porém, em meio a tantos problemas, Maria conhece a pequena Duda, sua vizinha; e o desajeitado Bob, que se tornam seus amigos. Bob é a materialização do príncipe de Maria em seus sonhos, e este a ajudará a conseguir emprego em sua lanchonete e a transformará na estrela de um show.',
    popularity: 6.416,
    poster_path: '/dkyfy7ifoYky2IeDCwVukOZmTwV.jpg',
    release_date: '1990-06-22',
    title: 'Xuxa - Lua de Cristal',
    vote_average: 5.5
  },
  argTypes: {
    genre_ids: { type: 'string' }
  }
} as Meta

export const Default: Story<MovieCardProps> = (args) => <MovieCard {...args} />
