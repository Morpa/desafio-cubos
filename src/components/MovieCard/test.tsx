import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import MovieCard from '.'

const props = {
  id: '1',
  genres: ['Suspense', 'Terror', 'Fantasia'],
  original_title: 'Xuxa - Lua de Cristal',
  overview:
    'Maria da Graça é uma jovem bonita e sonhadora que se muda para a cidade grande com a intenção de fazer aulas de canto. Lá, ela se hospeda na casa de sua tia Zuleika e seus primos Lidinha e Mauricinho, que vivem atormentando sua vida, fazendo-a trabalhar como uma escrava. Por ser um tanto ingênua e tímida, Maria vive caindo nas armações de Lidinha; enquanto é alvo das constantes cantadas de Mauricinho, ao que ela rejeita; tendo de suportar também os detestáveis amigos deste, enquanto não consegue se adaptar á cidade. Porém, em meio a tantos problemas, Maria conhece a pequena Duda, sua vizinha; e o desajeitado Bob, que se tornam seus amigos. Bob é a materialização do príncipe de Maria em seus sonhos, e este a ajudará a conseguir emprego em sua lanchonete e a transformará na estrela de um show.',
  popularity: 6.416,
  poster_path: 'dkyfy7ifoYky2IeDCwVukOZmTwV.jpg',
  release_date: '1990-06-22',
  title: 'Xuxa - Lua de Cristal',
  vote_average: 5.5
}

describe('<MovieCard />', () => {
  it('should render the correctly', () => {
    const { container } = renderWithTheme(<MovieCard {...props} />)

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('img', { name: `Poster of the movie ${props.title}` })
    ).toHaveAttribute(
      'src',
      `https://image.tmdb.org/t/p/w342/${props.poster_path}`
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
