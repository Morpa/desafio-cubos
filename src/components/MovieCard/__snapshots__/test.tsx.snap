// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> should render the correctly 1`] = `
.c5 {
  background: #116193;
  padding: 0.2rem;
  border-radius: 50%;
  color: #00ede7;
  position: relative;
  height: 5.6rem;
  width: 5.6rem;
}

.c5 .c6 {
  border: 0.3rem solid #00ede7;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  width: 100%;
  height: 100%;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.2rem 0.9rem;
  text-align: center;
  background: #FFF;
  border: 2px solid #116193;
  border-radius: 2rem;
  color: #116193;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
}

.c14:hover {
  background: #e5e5e5;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border-radius: 0.4rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  margin: 2rem 0;
  width: 100%;
  height: 32rem;
  margin-bottom: 4rem;
  background: #e8e8e8;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #5e5e5e;
}

.c2 {
  width: 20rem;
  height: 100%;
  background: #116193;
  object-fit: cover;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: block;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  justify-items: flex-start;
  height: 6.4rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding-left: 9.6rem;
  padding-right: 0.8rem;
  background: #116193;
  color: #00ede7;
  overflow: auto;
  white-space: nowrap;
  font-family: 'Abel';
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 2.4rem;
  overflow: hidden;
}

.c10 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  font-size: 1.8rem;
  color: #5e5e5e;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-height: 13rem;
  overflow: hidden;
  text-overflow: clip;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-top: 2.4rem;
  padding: 0.8rem 0;
}

.c12 .c13 {
  margin-right: 0.4rem;
}

.c4 {
  width: 7rem;
  height: 7rem;
  position: absolute;
  margin-top: 2.4rem;
  margin-left: 1.6rem;
}

@media (max-width:768px) {
  .c1 {
    -webkit-flex-direction: column-reverse;
    -ms-flex-direction: column-reverse;
    flex-direction: column-reverse;
    width: 100%;
    height: 100%;
  }
}

@media (max-width:768px) {
  .c2 {
    width: 95%;
    margin: 0 auto;
    max-height: 100%;
  }
}

@media (max-width:768px) {
  .c8 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    white-space: normal;
    min-height: 90%;
    font-size: 1.6rem;
  }
}

@media (max-width:768px) {
  .c12 {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    margin-top: 0.8rem;
  }

  .c12 .c13 {
    margin: 0.4rem;
  }
}

@media (max-width:768px) {
  .c4 {
    margin-top: 2.4rem;
  }
}

<div
  class="c0"
>
  <a
    class="c1"
    href="/movie/1"
  >
    <img
      alt="Poster of the movie Xuxa - Lua de Cristal"
      class="c2"
      src="https://image.tmdb.org/t/p/w342/dkyfy7ifoYky2IeDCwVukOZmTwV.jpg"
    />
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6 c7"
          >
            55
            %
          </div>
        </div>
      </div>
      <h1
        class="c8"
      >
        Xuxa - Lua de Cristal
      </h1>
      <div
        class="c9"
      >
        <p
          class="c10"
        >
          22/06/1990
        </p>
        <div
          class="c11"
        >
          Maria da Graça é uma jovem bonita e sonhadora que se muda para a cidade grande com a intenção de fazer aulas de canto. Lá, ela se hospeda na casa de sua tia Zuleika e seus primos Lidinha e Mauricinho, que vivem atormentando sua vida, fazendo-a trabalhar como uma escrava. Por ser um tanto ingênua e tímida, Maria vive caindo nas armações de Lidinha; enquanto é alvo das constantes cantadas de Mauricinho, ao que ela rejeita; tendo de suportar também os detestáveis amigos deste, enquanto não consegue se adaptar á cidade. Porém, em meio a tantos problemas, Maria conhece a pequena Duda, sua vizinha; e o desajeitado Bob, que se tornam seus amigos. Bob é a materialização do príncipe de Maria em seus sonhos, e este a ajudará a conseguir emprego em sua lanchonete e a transformará na estrela de um show.
        </div>
        <div
          class="c12"
        >
          <span
            class="c13 c14"
          >
            Suspense
          </span>
          <span
            class="c13 c14"
          >
            Terror
          </span>
          <span
            class="c13 c14"
          >
            Fantasia
          </span>
        </div>
      </div>
    </div>
  </a>
</div>
`;
